0004-post-tags.md

```diff
@@ -0,0 +1,324 @@
+
+For example a lot of communities offer mixed content such as Memes, News and Discussions. Some people would like to only see News or Discussions. In such a case they could blocklist (filter out) the "Meme" tag, going forward they would not be presented with any Memes in that community.
+
+Another use for tags is better handling of how posts are displayed. For example some people would like the content of a post to stay hidden for all users. This could be the case in a community focused on release discussions of Books, Movies, TV Shows or communities with content some users might find offensive but which is not neccesarily NSFW (for example jokes about certain topics).
+
+In order to ensure this consistency while filtering tags can only be created by privileged users, initially that is intended to be Admins and Community moderators.
+
+Essentially the tags serve as an easier way to message what is *inside* the post. Combining multiple tags allows for even finer control. Made a post about news in poland? Add the "News" and "Poland" tags to it and people interested exclusively in polish news can find it easily or people not interested can avoid it just as easily.
+
+# Reference-level explanation
+
+### Protocol:
+According to https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tags a general tag object exists in the ActivityPub protocol allowing for custom implementations so long as a type is specified.
+To keep things simple an initial community tag object should consist of the following:
+```json
+{
+  id: number,
+  url: string,
+  name: string
+}
+```
+
+Entirely unmoderated tags are not an option for lemmy as the moderation workload would be too much. Additionally users being able to type out tags themselves introduces splintering in the tag contents due to typos. A better solution is a curated list of tags users can attach to their posts. The list of tags can be maintained by both admins and moderators allowing for each community to tailor tags to their specific needs.
+Content for tags would be located in the respective root: `https://example.org/c/example/t/tag`
+
+The tag URL could then be utilized as a unique identifier for the tag, however doing so would restrict the ability to move the underlying tag url in the future. Instead an additional `id` field should be used. Using the object ID optional tag federation can also be achieved, allowing for communities across multiple instances to share content via tags (example: News tag shared across instances). This would also solve the issue of splintered communities across instances while not forcing it on the communities in question. For now tags will only be applicable to posts, however the general design allows for them to be attached to any kind of object later on, be that instance, community or user. The limited initial scope allows for easier modifications should any rough edges or missing features be discovered.
```

Comment:
* IDs in AP _must_ be URLs and globally referenceable. The URL should represent the ownership of the object and be resolveable. Otherwise, how would an instance know who to trust wrt updating / adding tags?
* Using a URL of the format `https://instance.com/tags/<UUID>` doesn't make sense because then the ID is still instance-local so doesn't help with community migration.
* Using a URL of the format `https://join-lemmy.org/tags/UUID` also doesn't make sense because then the tag object cannot be fetched and ownership is unclear.
* Making ids opaque is a neat idea, but due to the above doesn't solve many problems and it adds more complexity. Even if the id is hand-picked, we can still allow updates of the other properties of the tag (e.g. shown name and hierarchy).

```suggestion
The ID of a CommunityPostTag is a URL prefixed with `https://instance/c/community/tags/` and a user-chosen or generated (from the name) ID. The ID itself will rarely be user-visible, since the name is used there.

For now tags will only be applicable to posts, however the general design allows for them to be attached to any kind of object later on, be that instance, community or user. The limited initial scope allows for easier modifications should any rough edges or missing features be discovered.
```